File: Tests/EncodingTests.pas
namespace GoldLibrary.Tests.Shared;

uses
  RemObjects.Elements.EUnit,
  RemObjects.Elements.RTL;

type
  EncodingTests = public class(Test)
  public
    method Base64Test;
    begin
      Assert.AreEqual(TestApplication2.DoEncodeBase64(), 'YWJjMTIzIT8kKiYoKSctPUB+');
      Assert.AreEqual(TestApplication2.DoDecodeBase64(), "abc123!?$*&()'-=@~");
    end;

    method BinaryReadTest;
    begin
      Assert.AreEqual(TestApplication2.DoBinaryRead(), 3.141592653589793);
    end;

    method BinaryPutVaruintTest;
    begin
      var lBuf := TestApplication2.DoBinaryPutUvarint(1);
      Assert.AreEqual(lBuf.Length, 1);
      Assert.AreEqual(lBuf[0], 1);

      lBuf := TestApplication2.DoBinaryPutUvarint(127);
      Assert.AreEqual(lBuf.Length, 1);
      Assert.AreEqual(lBuf[0], $7f);

      lBuf := TestApplication2.DoBinaryPutUvarint(256);
      Assert.AreEqual(lBuf.Length, 2);
      Assert.AreEqual(lBuf[0], $80);
      Assert.AreEqual(lBuf[1], $02);
    end;

    method HexTest;
    begin
      Assert.AreEqual(TestApplication2.DoEncodingHexDecode(), 'Hello Gopher!');

      var lBuf := TestApplication2.DoEncodingHexEncode();
      Assert.AreEqual(lBuf.Length, 26);
      Assert.AreEqual(lBuf[0], 52);
      Assert.AreEqual(lBuf[1], 56);
    end;

    method JsonTest;
    begin
      var lString := TestApplication2.DoEncodingJsonMarshal();
      Assert.AreEqual(lString, '{"ID":1,"Name":"Reds","Colors":["Crimson","Red","Ruby","Maroon"]}'); // T81705 issue, T81774
    end;

    method PemTest;
    begin
      var lString: RemObjects.Elements.RTL.String := TestApplication2.DoEncodingPEMEncode();
      var lLines := lString.Split(#10, true);
      Assert.AreEqual(lLines.Count, 4);
      Assert.AreEqual(lLines[0], '-----BEGIN MESSAGE-----');
      Assert.AreEqual(lLines[3], '-----END MESSAGE-----');
      Assert.IsTrue(TestApplication2.DoDecodingPEMDecode());
    end;

  end;

end.

File: Tests/EncodingTests.go
package GoldLibrary.Tests.Shared

import "fmt"
import b64 "encoding/base64"
import "encoding/binary"
import "encoding/hex"
import "encoding/json"
import "encoding/pem"
import "crypto/x509"

func DoEncodeBase64() string {
	data := "abc123!?$*&()'-=@~"
	sEnc := b64.StdEncoding.EncodeToString([]byte(data))
	return sEnc
}

func DoDecodeBase64() string {
	data := "abc123!?$*&()'-=@~"
	sEnc := b64.StdEncoding.EncodeToString([]byte(data))

	sDec, _ := b64.StdEncoding.DecodeString(sEnc)
	return string(sDec)
}

func DoBinaryRead() float64 {
	var pi float64
	b := []byte{0x18, 0x2d, 0x44, 0x54, 0xfb, 0x21, 0x09, 0x40}
	buf := bytes.NewReader(b)
	err := binary.Read(buf, binary.LittleEndian, &pi)
	if err != nil {
		fmt.Println("binary.Read failed:", err)
	}
	return pi
}

func DoBinaryPutUvarint(data uint64) []byte {
	buf := make([]byte, binary.MaxVarintLen64)
	n := binary.PutUvarint(buf, data)
	return buf[:n]
}

func DoEncodingHexDecode() string {
	const s = "48656c6c6f20476f7068657221"
	decoded, err := hex.DecodeString(s)
	if err != nil {
		log.Fatal(err)
	}

	return string(decoded)
}

func DoEncodingHexEncode() []byte {
	src := []byte("Hello Gopher!")

	dst := make([]byte, hex.EncodedLen(len(src)))
	hex.Encode(dst, src)
	return dst
}

func DoEncodingJsonMarshal() string {

	type ColorGroup struct {
		ID     int
		Name   string
		Colors []string
	}
	group := ColorGroup{
		ID:     1,
		Name:   "Reds",
		Colors: []string{"Crimson", "Red", "Ruby", "Maroon"},
	}
	b, err := json.Marshal(group)
	if err != nil {
		fmt.Println("error:", err)
	}

	return string(b)
}

func DoEncodingPEMEncode() string {
	block := &pem.Block{
		Type: "MESSAGE",
		Headers: map[string]string{
			"Animal": "Gopher",
		},
		Bytes: []byte("test"),
	}

	lBytes := pem.EncodeToMemory(block)
	return string(lBytes)
}

func DoDecodingPEMDecode() Boolean {
	var pubPEMData = []byte(`
-----BEGIN PUBLIC KEY-----
MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAlRuRnThUjU8/prwYxbty
WPT9pURI3lbsKMiB6Fn/VHOKE13p4D8xgOCADpdRagdT6n4etr9atzDKUSvpMtR3
CP5noNc97WiNCggBjVWhs7szEe8ugyqF23XwpHQ6uV1LKH50m92MbOWfCtjU9p/x
qhNpQQ1AZhqNy5Gevap5k8XzRmjSldNAFZMY7Yv3Gi+nyCwGwpVtBUwhuLzgNFK/
yDtw2WcWmUU7NuC8Q6MWvPebxVtCfVp/iQU6q60yyt6aGOBkhAX0LpKAEhKidixY
nP9PNVBvxgu3XZ4P36gZV6+ummKdBVnc3NqwBLu5+CcdRdusmHPHd5pHf4/38Z3/
6qU2a/fPvWzceVTEgZ47QjFMTCTmCwNt29cvi7zZeQzjtwQgn4ipN9NibRH/Ax/q
TbIzHfrJ1xa2RteWSdFjwtxi9C20HUkjXSeI4YlzQMH0fPX6KCE7aVePTOnB69I/
a9/q96DiXZajwlpq3wFctrs1oXqBp5DVrCIj8hU2wNgB7LtQ1mCtsYz//heai0K9
PhE4X6hiE0YmeAZjR0uHl8M/5aW9xCoJ72+12kKpWAa0SFRWLy6FejNYCYpkupVJ
yecLk/4L1W0l6jQQZnWErXZYe0PNFcmwGXy1Rep83kfBRNKRy5tvocalLlwXLdUk
AIU+2GKjyT3iMuzZxxFxPFMCAwEAAQ==
-----END PUBLIC KEY-----
and some more`)

block, rest := pem.Decode(pubPEMData)
if block == nil || block.Type != "PUBLIC KEY" {
	log.Fatal("failed to decode PEM block containing public key")
}

pub, err := x509.ParsePKIXPublicKey(block.Bytes)
if err != nil {
	log.Fatal(err)
}

return string(rest) == "and some more"
//fmt.Printf("Got a %T, with remaining data: %q", pub, rest)
}